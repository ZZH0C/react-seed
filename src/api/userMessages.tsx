import axios from 'axios';

async function getMessage(id: string, token: string | null) {
  const url = `https://gmail.googleapis.com/gmail/v1/users/me/messages/${id}`;

  try {
    const response = await axios.get(url, {
      headers: { Authorization: `Bearer ${token}` },
    });
    return response.data;
  } catch (error) {
    //TODO: write nice error handler
    console.error(error);
  }
}

async function getMessageList(token: string | null) {
  const url =
    'https://gmail.googleapis.com/gmail/v1/users/me/messages?maxResults=3';
  try {
    const response = await axios.get(url, {
      headers: { Authorization: `Bearer ${token}` },
    });
    return response.data;
  } catch (error) {
    //TODO: write nice error handler
    console.error(error);
  }
}

export async function loadMessages(token: string | null) {
  const idList = await getMessageList(token);
  const messageList = idList.messages.map(async (e: { id: string }) => {
    return await getMessage(e.id, token);
  });
  return await Promise.allSettled(messageList);
}
