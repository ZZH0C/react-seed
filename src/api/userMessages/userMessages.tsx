import axios from 'axios';

const getMessage = async (id: string, token: string) => {
  const url = `https://gmail.googleapis.com/gmail/v1/users/me/messages/${id}`;
  try {
    const response = await axios.get(url, {
      headers: { Authorization: `Bearer ${token}` },
    });
    return response.data;
  } catch (error) {
    //TODO: write nice error handler
    console.error(error);
  }
};

const getMessageList = async (token: string, category: string) => {
  const url = `https://gmail.googleapis.com/gmail/v1/users/me/messages?maxResults=3${category}`;
  try {
    const response = await axios.get(url, {
      headers: { Authorization: `Bearer ${token}` },
    });
    return response.data;
  } catch (error) {
    //TODO: write nice error handler
    console.error(error);
  }
};

export const loadMessages = async (token: string, category: string) => {
  const idList = await getMessageList(token, category);
  if (idList) {
    const messageList = idList.messages.map(async (e: { id: string }) => {
      return await getMessage(e.id, token);
    });
    return await Promise.allSettled(messageList);
  } else return [];
};

export const loadOneMessage = async (id: string, token: string) => {
  return await getMessage(id, token);
};
