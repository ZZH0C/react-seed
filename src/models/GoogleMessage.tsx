export interface GoogleMessage {
  value: {
    snippet: string;
    payload: { headers: { name: string; value: string }[] };
  };
}
