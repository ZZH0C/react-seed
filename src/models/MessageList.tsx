export interface MessageList {
  nextPageToken: any;
  list: Array<
    PromiseSettledResult<
      Promise<any> extends PromiseLike<infer U> ? U : Promise<any>
    >
  >;
}
