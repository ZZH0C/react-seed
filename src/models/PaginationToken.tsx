export interface PaginationToken {
  currentPage: number;
  pageTokens: string[];
}
