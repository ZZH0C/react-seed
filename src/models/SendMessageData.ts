export interface sendMessageData {
  to?: string;
  cc?: string;
  bcc?: string;
  subject?: string;
  messageText?: string;
}
