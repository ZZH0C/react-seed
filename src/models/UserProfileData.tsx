export interface UserProfileData {
  googleId: string;
  imageUrl: string;
  email: string;
  name: string;
  givenName: string;
  familyName: string;
}
