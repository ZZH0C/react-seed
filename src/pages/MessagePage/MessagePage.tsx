import React, { useContext, useEffect, useState } from 'react';
import { Link, useLocation } from 'react-router-dom';
import { UserContext } from '../../App';
import { loadOneMessage } from '../../api/userMessages/userMessages';
import { sortMessageData } from '../../hooks/useCreateMessagesUi';

// const sortMessage = (data) => {
//
// }

export const MessagePage: React.FC = () => {
  const userData = useContext(UserContext);
  const [state, setState] = useState<any>();
  const location = useLocation();
  const messageId: any = location.state;
  useEffect(() => {
    if (userData && 'accessToken' in userData) {
      loadOneMessage(messageId, userData.accessToken).then((r: any) =>
        setState(r),
      );
    }
  }, []);
  let text: {
    snippet: any;
    date: string;
    from: string;
    text: any;
    title: string;
  } = {
    from: '',
    snippet: '',
    title: '',
    date: '',
    text: '',
  };
  if (state) text = sortMessageData(state);
  if (text.text === '') {
    text.text = text.snippet;
  }
  return (
    <>
      <Link to="/home">
        <button className={'button_home'}>Return Home</button>
      </Link>
      <section className={'message_section_container'}>
        <div className={'message_section'}>
          <span>From: {text.from}</span>
          <span>Subject: {text.title}</span>
          <span> {text.text}</span>
        </div>
      </section>
    </>
  );
};
