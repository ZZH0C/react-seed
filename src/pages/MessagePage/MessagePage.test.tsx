import { configure, shallow } from 'enzyme';
import { MessagePage } from './MessagePage';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
import { loadOneMessage } from '../../api/userMessages/userMessages';
import { mocked } from 'ts-jest/utils';
import { GoogleMessage } from '../../models/GoogleMessage';
import { useLocation } from 'react-router-dom';

const mockResponse: GoogleMessage = {
  value: {} as GoogleMessage,
  id: 'SOME_ID',
  snippet: 'SOME_SNIPPET',
  payload: {
    headers: [{ name: 'SOME_NAME', value: 'SOME_VALUE' }],
    parts: [
      {
        body: { data: 'SOME_DATA' },
      },
    ],
  },
};
// jest.mock('react-router-dom');
//
// const useLocation = jest.fn(() => ({
//   pathname: '',
//   search: '',
//   hash: '',
//   state: {
//     id: 'someId',
//     token: 'someToken',
//   },
// }));

mocked(useLocation).mockImplementation(() => ({
  pathname: '',
  search: '',
  hash: '',
  state: {
    id: 'someId',
    token: 'someToken',
  },
}));

jest.mock('../../api/userMessages/userMessages');
mocked(loadOneMessage).mockImplementation(() => {
  return Promise.resolve(mockResponse);
});
// mocked(useLocation).mockImplementation(() => {
//   return {
//     pathname: '',
//     search: '',
//     hash: '',
//     state: {
//       id: 'someId',
//       token: 'someToken',
//     },
//   };
// });

describe('components/MessagePage', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render', () => {
    expect(shallow(<MessagePage />)).toMatchSnapshot();
    expect(useLocation).toBeCalledTimes(1);
  });

  // TODO:
  // it('should render and call loadOneMessage', () => {
  //   shallow(<MessagePage />);
  // expect(loadOneMessage).toBeCalledTimes(1);
  // });
});
