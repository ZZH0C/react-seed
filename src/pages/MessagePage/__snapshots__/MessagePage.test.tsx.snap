// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/MessagePage should render 1`] = `
<Fragment>
  <ButtonLink
    href="/"
  >
    Return Home
  </ButtonLink>
  <Message
    from=""
    title=""
  />
</Fragment>
`;
